File: src/com/jetbrains/idear/nlp/NlpParserService.java
package com.jetbrains.idear.nlp;

import com.intellij.ide.plugins.IdeaPluginDescriptor;
import com.intellij.ide.plugins.PluginManager;
import com.intellij.openapi.extensions.PluginId;
import opennlp.tools.cmdline.parser.ParserTool;
import opennlp.tools.parser.Parse;
import opennlp.tools.parser.Parser;
import opennlp.tools.parser.ParserFactory;
import opennlp.tools.parser.ParserModel;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.TestOnly;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;

public class NlpParserService extends ParserService {
    private String path;
    private Parser parser;

    @TestOnly
    public NlpParserService(String testPath) {
        this.path = testPath;
    }

    public NlpParserService() {
    }

    @Nullable
    public Parse parseSentence(String sentence) {
        if (parser == null) {
            throw new IllegalStateException();
        }
        Parse topParses[] = ParserTool.parseLine(sentence, parser, 1);
        assert topParses.length == 1;
        return topParses[0];
    }

    public void init() {
        ParserModel model = readParserModel();
        if (model == null) return;
        parser = ParserFactory.create(model);
    }

    @Nullable
    private ParserModel readParserModel() {
        ParserModel model = null;
        try {
            model = new ParserModel(getModelInputStream());
            getModelInputStream().close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return model;
    }

    private InputStream getModelInputStream() throws FileNotFoundException {
        if (path != null) {
            return new FileInputStream(path);
        }

        PluginId id = PluginId.getId("com.jetbrains.idear");
        IdeaPluginDescriptor plugin = PluginManager.getPlugin(id);

        assert plugin != null;

        ClassLoader classLoader = plugin.getPluginClassLoader();
        return classLoader.getResourceAsStream("en-parser-chunking.bin");
    }

}

File: src/main/java/org/openasr/idear/nlp/NlpParserService.kt
package org.openasr.idear.nlp

import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.extensions.PluginId
import opennlp.tools.cmdline.parser.ParserTool
import opennlp.tools.parser.*
import org.jetbrains.annotations.TestOnly
import java.io.*

class NlpParserService @TestOnly constructor(path: String?) : ParserService() {
    private lateinit var parser: Parser
    private val logger = Logger.getInstance(javaClass)
    val plugin by lazy { PluginManagerCore.getPlugin(PluginId.getId("com.jetbrains.idear"))!! }

    private val modelInputStream: InputStream =
            if (path != null) FileInputStream(path)
            else plugin.pluginClassLoader.getResourceAsStream("en-parser-chunking.bin")

    override fun parseSentence(sentence: String): Parse? =
            ParserTool.parseLine(sentence, parser, 1).apply { assert(size == 1) }[0]

    override fun init() {
        val model = readParserModel() ?: return
        parser = ParserFactory.create(model)
    }

    private fun readParserModel() =
            try {
                val model = ParserModel(modelInputStream)
                modelInputStream.close()
                model
            } catch (ioException: IOException) {
                logger.error(ioException)
                null
            }
}